proposals/csharp-10.0/lambda-improvements.md

```diff
@@ -0,0 +1,121 @@
+Support lambdas with attributes, explicit return type, and natural type.
+
+## Motivation
+Support for attributes on lambdas would provide parity with methods and local functions.
+
+Support for explicit return types would provide symmetry with lambda parameters where explicit types can be specified.
+Allowing explicit return types would also provide control over compiler performance in nested lambdas where overload resolution must bind the lambda body currently to determine the signature.
+
+A natural type for lambda expressions and method groups will allow more scenarios where lambdas and method groups may be used without an explicit delegate type, including as initializers in `var` declarations.
+Requiring explicit delegate types for lambdas and method groups has been a friction point for customers, and has become an impediment to progress in ASP.NET with recent work on [MapAction](https://github.com/dotnet/aspnetcore/pull/29878).
+
+## Attributes
+Attributes may be added to lambda expressions.
+```csharp
+f = [MyAttribute] x => x;          // [MyAttribute]lambda
+f = [MyAttribute] (int x) => x;    // [MyAttribute]lambda
+f = [MyAttribute] static x => x;   // [MyAttribute]lambda
+f = [return: MyAttribute] () => 1; // [return: MyAttribute]lambda
+```
+Attributes may be added to lambda parameters that are declared with explicit types.
+```csharp
+f = ([MyAttribute] x) => x;      // syntax error
+f = ([MyAttribute] int x) => x;  // [MyAttribute]x
+```
+
+Attributes are not supported for anonymous methods declared with `delegate { }` syntax.
+```csharp
+f = [MyAttribute] delegate { return 1; };         // syntax error
+f = delegate ([MyAttribute] int x) { return x; }; // syntax error
+```
+
+Attributes on the lambda or lambda parameters will be emitted to metadata on the method that maps to the lambda. (If the lambda does not require a closure class, the lambda is emitted as a method on the containing type. Otherwise the lambda is as a method on the generated closure class.)
+
+## Explicit return type
+An explicit return type may be specified after the parameter list.
+```csharp
+f = () : T => default;              // () : T
+f = x : int => 1;                   // <unknown> : int
+f = (ref int x) : ref int => ref x; // ref int : ref int
+f = static _ : void => { };         // <unknown> : void
+```
+
+Explicit return types are not supported for anonymous methods declared with `delegate { }` syntax.
+```csharp
+f = delegate : int { return 1; };         // syntax error
+f = delegate (int x) : int { return x; }; // syntax error
+```
+
+## Natural type
+A lambda expression has a natural type if the parameters types are explicit and either the return type is explicit or there is a common type from the natural types of all `return` expressions in the body. Otherwise there is no natural type.
+
+A method group has a natural type if the method group contains a single method and the method or reduced extension method has no unbound type parameters.
```

Comment:
Do we need to explicitly mention reduced extension methods? I think just "the method" would be fine?